File: amplify/functions/responses-api/handler.ts
import type { Handler } from 'aws-lambda';
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import {
  DynamoDBDocumentClient,
  QueryCommand,
  ScanCommand,
  GetCommand,
  UpdateCommand,
} from '@aws-sdk/lib-dynamodb';
import {
  SecretsManagerClient,
  GetSecretValueCommand,
} from '@aws-sdk/client-secrets-manager';
import { timingSafeEqual } from 'crypto';

const dynamodbClient = new DynamoDBClient({});
const dynamodb = DynamoDBDocumentClient.from(dynamodbClient);
const secretsClient = new SecretsManagerClient({});

// Cache the API key to avoid fetching from Secrets Manager on every request
let cachedApiKey: string | null = null;
let cacheExpiry = 0;
const CACHE_TTL_MS = 5 * 60 * 1000; // 5 minutes

async function getApiKey(): Promise<string> {
  const now = Date.now();

  // Return cached value if still valid
  if (cachedApiKey && cacheExpiry > now) {
    return cachedApiKey;
  }

  const secretArn = process.env.API_KEY_SECRET_ARN;
  if (!secretArn) {
    throw new Error('API_KEY_SECRET_ARN not configured');
  }

  const result = await secretsClient.send(
    new GetSecretValueCommand({ SecretId: secretArn })
  );

  if (!result.SecretString) {
    throw new Error('Secret value is empty');
  }

  const secret = JSON.parse(result.SecretString);
  cachedApiKey = secret.apiKey;
  cacheExpiry = now + CACHE_TTL_MS;

  return cachedApiKey!;
}

// ============ SECURITY: Rate Limiting ============
const RATE_LIMIT = 100; // requests per minute per API key
const rateLimitMap = new Map<string, { count: number; resetAt: number }>();

function checkRateLimit(apiKey: string): boolean {
  const now = Date.now();
  const record = rateLimitMap.get(apiKey) || { count: 0, resetAt: now + 60000 };

  if (record.resetAt < now) {
    record.count = 1;
    record.resetAt = now + 60000;
  } else {
    record.count++;
  }

  rateLimitMap.set(apiKey, record);
  return record.count <= RATE_LIMIT;
}

// ============ SECURITY: Structured Logging ============
type LogLevel = 'INFO' | 'WARN' | 'ERROR' | 'AUDIT';

function log(level: LogLevel, message: string, data?: Record<string, unknown>) {
  console.log(
    JSON.stringify({
      level,
      timestamp: new Date().toISOString(),
      message,
      ...data,
    })
  );
}

// ============ Response Helpers ============
const CORS_HEADERS = {
  'Access-Control-Allow-Origin': process.env.ALLOWED_ORIGIN || '*',
  'Access-Control-Allow-Headers': 'Content-Type,x-api-key',
  'Access-Control-Allow-Methods': 'GET,POST,OPTIONS',
  'Content-Type': 'application/json',
};

interface LambdaResponse {
  statusCode: number;
  headers: Record<string, string>;
  body: string;
}

function response(statusCode: number, body: unknown): LambdaResponse {
  return {
    statusCode,
    headers: CORS_HEADERS,
    body: JSON.stringify(body),
  };
}

interface FunctionUrlEvent {
  requestContext: {
    http: {
      method: string;
      path: string;
    };
  };
  headers: Record<string, string | undefined>;
  body?: string;
}

// ============ SECURITY: Timing-Safe API Key Validation ============
async function validateApiKey(event: FunctionUrlEvent): Promise<boolean> {
  const apiKey = event.headers['x-api-key'] || event.headers['X-Api-Key'] || '';

  if (!apiKey) {
    return false;
  }

  try {
    const expectedKey = await getApiKey();

    if (!expectedKey) {
      return false;
    }

    // Prevent timing attacks by using constant-time comparison
    if (apiKey.length !== expectedKey.length) {
      return false;
    }

    return timingSafeEqual(Buffer.from(apiKey), Buffer.from(expectedKey));
  } catch (error) {
    log('ERROR', 'Failed to validate API key', {
      error: error instanceof Error ? error.message : 'Unknown error',
    });
    return false;
  }
}

// ============ Main Handler ============
export const handler: Handler<FunctionUrlEvent, LambdaResponse> = async (event) => {
  const method = event.requestContext?.http?.method || 'GET';
  const path = event.requestContext?.http?.path || '/';
  const apiKey = event.headers['x-api-key'] || event.headers['X-Api-Key'] || '';

  log('INFO', 'Request received', { method, path });

  // Health check endpoint (no auth required)
  if (method === 'GET' && path === '/health') {
    return response(200, {
      status: 'healthy',
      timestamp: new Date().toISOString(),
    });
  }

  // Handle CORS preflight
  if (method === 'OPTIONS') {
    return response(200, {});
  }

  // SECURITY: Validate API Key (now async)
  if (!(await validateApiKey(event))) {
    log('AUDIT', 'Authentication failed', { path, method });
    return response(401, { error: 'Unauthorized', message: 'Invalid or missing API key' });
  }

  // SECURITY: Check rate limit
  if (!checkRateLimit(apiKey)) {
    log('WARN', 'Rate limit exceeded', { apiKey: apiKey.substring(0, 8) + '...' });
    return response(429, { error: 'Too many requests', message: 'Rate limit exceeded' });
  }

  try {
    // GET /responses/pending - List pending responses
    if (method === 'GET' && path === '/responses/pending') {
      log('AUDIT', 'Listing pending responses');
      return await listPendingResponses();
    }

    // GET /responses/{id}/download - Download and mark as downloaded
    const downloadMatch = path.match(/^\/responses\/([^/]+)\/download$/);
    if (method === 'GET' && downloadMatch) {
      const id = downloadMatch[1];
      log('AUDIT', 'Downloading response', { id });
      return await downloadAndMarkResponse(id);
    }

    // POST /responses/{id}/unmark - Unmark for re-download
    const unmarkMatch = path.match(/^\/responses\/([^/]+)\/unmark$/);
    if (method === 'POST' && unmarkMatch) {
      const id = unmarkMatch[1];
      log('AUDIT', 'Unmarking response', { id });
      return await unmarkResponse(id);
    }

    // GET /responses/all - List all responses (restricted in production)
    if (method === 'GET' && path === '/responses/all') {
      log('AUDIT', 'Listing all responses');
      return await listAllResponses();
    }

    log('WARN', 'Route not found', { path, method });
    return response(404, { error: 'Not found' });
  } catch (error) {
    log('ERROR', 'Handler error', {
      error: error instanceof Error ? error.message : 'Unknown error',
      stack: error instanceof Error ? error.stack : undefined,
    });
    return response(500, {
      error: 'Internal server error',
    });
  }
};

async function listPendingResponses(): Promise<LambdaResponse> {
  const tableName = process.env.CUESTIONARIO_RESPONSE_TABLE_NAME;
  if (!tableName) {
    return response(500, { error: 'Configuration error' });
  }

  // Use the downloadStatus GSI for efficient query
  const indexName = process.env.DOWNLOAD_STATUS_INDEX_NAME;

  let result;
  if (indexName) {
    // Use GSI query (more efficient)
    result = await dynamodb.send(
      new QueryCommand({
        TableName: tableName,
        IndexName: indexName,
        KeyConditionExpression: 'downloadStatus = :status',
        ExpressionAttributeValues: { ':status': 'pending' },
      })
    );
  } else {
    // Fallback to scan with filter
    result = await dynamodb.send(
      new ScanCommand({
        TableName: tableName,
        FilterExpression: 'downloadStatus = :status',
        ExpressionAttributeValues: { ':status': 'pending' },
      })
    );
  }

  // Map to external format (hide internal details)
  const responses = (result.Items || []).map((item: Record<string, unknown>) => ({
    id: item.id,
    tokenId: item.tokenId,
    cuestionarioId: item.cuestionarioId,
    cuestionarioVersion: item.cuestionarioVersion,
    cuestionarioTitle: item.cuestionarioTitle,
    startedAt: item.startedAt,
    finishedAt: item.finishedAt,
    totalTimeMs: item.totalTimeMs,
    totalTimeAdjustedMs: item.totalTimeAdjustedMs,
    status: item.status,
    downloadStatus: item.downloadStatus,
    createdAt: item.createdAt,
  }));

  return response(200, {
    count: responses.length,
    responses,
  });
}

async function listAllResponses(): Promise<LambdaResponse> {
  const tableName = process.env.CUESTIONARIO_RESPONSE_TABLE_NAME;
  if (!tableName) {
    return response(500, { error: 'Configuration error' });
  }

  const result = await dynamodb.send(
    new ScanCommand({
      TableName: tableName,
    })
  );

  // Map to external format
  const responses = (result.Items || []).map((item: Record<string, unknown>) => ({
    id: item.id,
    tokenId: item.tokenId,
    cuestionarioId: item.cuestionarioId,
    cuestionarioVersion: item.cuestionarioVersion,
    cuestionarioTitle: item.cuestionarioTitle,
    startedAt: item.startedAt,
    finishedAt: item.finishedAt,
    totalTimeMs: item.totalTimeMs,
    totalTimeAdjustedMs: item.totalTimeAdjustedMs,
    status: item.status,
    downloadStatus: item.downloadStatus,
    downloadedAt: item.downloadedAt,
    downloadedBy: item.downloadedBy,
    createdAt: item.createdAt,
  }));

  return response(200, {
    count: responses.length,
    responses,
  });
}

async function downloadAndMarkResponse(id: string): Promise<LambdaResponse> {
  const tableName = process.env.CUESTIONARIO_RESPONSE_TABLE_NAME;

  if (!tableName) {
    return response(500, { error: 'Configuration error' });
  }

  // Get the response from DynamoDB
  const getResult = await dynamodb.send(
    new GetCommand({
      TableName: tableName,
      Key: { id },
    })
  );

  if (!getResult.Item) {
    return response(404, { error: 'Response not found' });
  }

  const record = getResult.Item;

  // Mark as downloaded in DynamoDB
  await dynamodb.send(
    new UpdateCommand({
      TableName: tableName,
      Key: { id },
      UpdateExpression:
        'SET downloadStatus = :status, downloadedAt = :downloadedAt, downloadedBy = :downloadedBy',
      ExpressionAttributeValues: {
        ':status': 'downloaded',
        ':downloadedAt': new Date().toISOString(),
        ':downloadedBy': 'external-api',
      },
    })
  );

  log('INFO', 'Response downloaded successfully', { id });

  // Return the full response data including answers
  return response(200, {
    id,
    downloadedAt: new Date().toISOString(),
    response: {
      tokenId: record.tokenId,
      cuestionarioId: record.cuestionarioId,
      cuestionarioVersion: record.cuestionarioVersion,
      cuestionarioTitle: record.cuestionarioTitle,
      startedAt: record.startedAt,
      finishedAt: record.finishedAt,
      totalTimeMs: record.totalTimeMs,
      totalTimeAdjustedMs: record.totalTimeAdjustedMs,
      status: record.status,
      answers: record.answersJson,
    },
  });
}

async function unmarkResponse(id: string): Promise<LambdaResponse> {
  const tableName = process.env.CUESTIONARIO_RESPONSE_TABLE_NAME;
  if (!tableName) {
    return response(500, { error: 'Configuration error' });
  }

  // Update status back to pending
  await dynamodb.send(
    new UpdateCommand({
      TableName: tableName,
      Key: { id },
      UpdateExpression:
        'SET downloadStatus = :status, downloadedAt = :downloadedAt, downloadedBy = :downloadedBy',
      ExpressionAttributeValues: {
        ':status': 'pending',
        ':downloadedAt': null,
        ':downloadedBy': null,
      },
    })
  );

  log('INFO', 'Response unmarked', { id });

  return response(200, {
    id,
    status: 'pending',
    message: 'Response unmarked for re-download',
  });
}

File: amplify/functions/responses-api/handler.py
import json
import os
import time
import hmac
import boto3
from datetime import datetime
from decimal import Decimal


# Initialize AWS clients
dynamodb = boto3.resource('dynamodb')
secrets_client = boto3.client('secretsmanager')

# Cache for API key
cached_api_key = None
cache_expiry = 0
CACHE_TTL = 300  # 5 minutes in seconds

# Rate limiting
rate_limits = {}
RATE_LIMIT = 100  # requests per minute


# ============ Helpers ============

def get_cors_headers():
    return {
        'Access-Control-Allow-Origin': os.environ.get('ALLOWED_ORIGIN', '*'),
        'Access-Control-Allow-Headers': 'Content-Type,x-api-key,X-Api-Key',
        'Access-Control-Allow-Methods': 'GET,POST,OPTIONS',
        'Content-Type': 'application/json'
    }


class DecimalEncoder(json.JSONEncoder):
    """Custom JSON encoder that handles Decimal types from DynamoDB"""
    def default(self, obj):
        if isinstance(obj, Decimal):
            # Convert to int if it's a whole number, otherwise float
            if obj % 1 == 0:
                return int(obj)
            return float(obj)
        return super().default(obj)


def response(status_code, body):
    return {
        'statusCode': status_code,
        'headers': get_cors_headers(),
        'body': json.dumps(body, cls=DecimalEncoder)
    }


def log(level, message, **kwargs):
    """Structured logging"""
    print(json.dumps({
        'level': level,
        'timestamp': datetime.utcnow().isoformat() + 'Z',
        'message': message,
        **kwargs
    }))


# ============ Security ============

def get_api_key():
    """Get API key from Secrets Manager with caching"""
    global cached_api_key, cache_expiry

    now = time.time()
    if cached_api_key and cache_expiry > now:
        return cached_api_key

    secret_arn = os.environ.get('API_KEY_SECRET_ARN')
    if not secret_arn:
        raise Exception('API_KEY_SECRET_ARN not configured')

    result = secrets_client.get_secret_value(SecretId=secret_arn)
    secret = json.loads(result['SecretString'])

    cached_api_key = secret.get('apiKey')
    cache_expiry = now + CACHE_TTL

    return cached_api_key


def validate_api_key(provided_key):
    """Validate API key using timing-safe comparison"""
    if not provided_key:
        return False

    try:
        expected_key = get_api_key()
        if not expected_key:
            return False

        # Timing-safe comparison to prevent timing attacks
        return hmac.compare_digest(provided_key, expected_key)
    except Exception as e:
        log('ERROR', 'Failed to validate API key', error=str(e))
        return False


def check_rate_limit(api_key):
    """Simple in-memory rate limiting"""
    global rate_limits

    now = time.time()
    key_prefix = api_key[:8] if api_key else 'unknown'

    record = rate_limits.get(key_prefix, {'count': 0, 'reset_at': now + 60})

    if record['reset_at'] < now:
        record = {'count': 1, 'reset_at': now + 60}
    else:
        record['count'] += 1

    rate_limits[key_prefix] = record
    return record['count'] <= RATE_LIMIT


# ============ Route Handlers ============

def list_pending_responses():
    """List responses with downloadStatus = pending using GSI"""
    table_name = os.environ.get('CUESTIONARIO_RESPONSE_TABLE_NAME')
    if not table_name:
        return response(500, {'error': 'Configuration error'})

    table = dynamodb.Table(table_name)

    # Use GSI for efficient query instead of scan
    # GSI name follows Amplify Gen 2 convention: {model}ByDownloadStatus
    gsi_name = os.environ.get('DOWNLOAD_STATUS_GSI_NAME', 'CuestionarioResponseByDownloadStatus')

    try:
        result = table.query(
            IndexName=gsi_name,
            KeyConditionExpression='downloadStatus = :status',
            ExpressionAttributeValues={':status': 'pending'}
        )
    except Exception as e:
        # Fallback to scan if GSI doesn't exist
        log('WARN', 'GSI query failed, falling back to scan', error=str(e), gsi=gsi_name)
        result = table.scan(
            FilterExpression='downloadStatus = :status',
            ExpressionAttributeValues={':status': 'pending'}
        )

    # Map to external format
    responses_list = []
    for item in result.get('Items', []):
        responses_list.append({
            'id': item.get('id'),
            'tokenId': item.get('tokenId'),
            'cuestionarioId': item.get('cuestionarioId'),
            'cuestionarioVersion': item.get('cuestionarioVersion'),
            'cuestionarioTitle': item.get('cuestionarioTitle'),
            'startedAt': item.get('startedAt'),
            'finishedAt': item.get('finishedAt'),
            'totalTimeMs': item.get('totalTimeMs'),
            'totalTimeAdjustedMs': item.get('totalTimeAdjustedMs'),
            'status': item.get('status'),
            'downloadStatus': item.get('downloadStatus'),
            'createdAt': item.get('createdAt'),
        })

    return response(200, {
        'count': len(responses_list),
        'responses': responses_list
    })


def list_all_responses():
    """List all responses"""
    table_name = os.environ.get('CUESTIONARIO_RESPONSE_TABLE_NAME')
    if not table_name:
        return response(500, {'error': 'Configuration error'})

    table = dynamodb.Table(table_name)
    result = table.scan()

    # Map to external format
    responses_list = []
    for item in result.get('Items', []):
        responses_list.append({
            'id': item.get('id'),
            'tokenId': item.get('tokenId'),
            'cuestionarioId': item.get('cuestionarioId'),
            'cuestionarioVersion': item.get('cuestionarioVersion'),
            'cuestionarioTitle': item.get('cuestionarioTitle'),
            'startedAt': item.get('startedAt'),
            'finishedAt': item.get('finishedAt'),
            'totalTimeMs': item.get('totalTimeMs'),
            'totalTimeAdjustedMs': item.get('totalTimeAdjustedMs'),
            'status': item.get('status'),
            'downloadStatus': item.get('downloadStatus'),
            'downloadedAt': item.get('downloadedAt'),
            'downloadedBy': item.get('downloadedBy'),
            'createdAt': item.get('createdAt'),
        })

    return response(200, {
        'count': len(responses_list),
        'responses': responses_list
    })


def download_and_mark_response(response_id):
    """Download response and mark as downloaded"""
    table_name = os.environ.get('CUESTIONARIO_RESPONSE_TABLE_NAME')
    if not table_name:
        return response(500, {'error': 'Configuration error'})

    table = dynamodb.Table(table_name)

    # Get the response
    result = table.get_item(Key={'id': response_id})
    item = result.get('Item')

    if not item:
        return response(404, {'error': 'Response not found'})

    # Mark as downloaded
    now = datetime.utcnow().isoformat() + 'Z'
    table.update_item(
        Key={'id': response_id},
        UpdateExpression='SET downloadStatus = :status, downloadedAt = :downloadedAt, downloadedBy = :downloadedBy',
        ExpressionAttributeValues={
            ':status': 'downloaded',
            ':downloadedAt': now,
            ':downloadedBy': 'external-api'
        }
    )

    log('INFO', 'Response downloaded successfully', id=response_id)

    # Return full response data
    return response(200, {
        'id': response_id,
        'downloadedAt': now,
        'response': {
            'tokenId': item.get('tokenId'),
            'cuestionarioId': item.get('cuestionarioId'),
            'cuestionarioVersion': item.get('cuestionarioVersion'),
            'cuestionarioTitle': item.get('cuestionarioTitle'),
            'startedAt': item.get('startedAt'),
            'finishedAt': item.get('finishedAt'),
            'totalTimeMs': item.get('totalTimeMs'),
            'totalTimeAdjustedMs': item.get('totalTimeAdjustedMs'),
            'status': item.get('status'),
            'answers': item.get('answersJson'),
        }
    })


def unmark_response(response_id):
    """Unmark response for re-download"""
    table_name = os.environ.get('CUESTIONARIO_RESPONSE_TABLE_NAME')
    if not table_name:
        return response(500, {'error': 'Configuration error'})

    table = dynamodb.Table(table_name)

    # Reset download status
    table.update_item(
        Key={'id': response_id},
        UpdateExpression='SET downloadStatus = :status REMOVE downloadedAt, downloadedBy',
        ExpressionAttributeValues={
            ':status': 'pending'
        }
    )

    log('INFO', 'Response unmarked', id=response_id)

    return response(200, {
        'id': response_id,
        'status': 'pending',
        'message': 'Response unmarked for re-download'
    })


# ============ Main Handler ============

def handler(event, context):
    """Main Lambda handler"""
    method = event.get('requestContext', {}).get('http', {}).get('method', 'GET')
    path = event.get('requestContext', {}).get('http', {}).get('path', '/')
    headers = event.get('headers', {})

    log('INFO', 'Request received', method=method, path=path)

    # Health check (no auth required)
    if method == 'GET' and path == '/health':
        return response(200, {
            'status': 'healthy',
            'timestamp': datetime.utcnow().isoformat() + 'Z'
        })

    # Handle CORS preflight
    if method == 'OPTIONS':
        return response(200, {})

    # Get API key from headers (case-insensitive)
    api_key = headers.get('x-api-key') or headers.get('X-Api-Key') or ''

    # Validate API key
    if not validate_api_key(api_key):
        log('AUDIT', 'Authentication failed', path=path, method=method)
        return response(401, {'error': 'Unauthorized', 'message': 'Invalid or missing API key'})

    # Check rate limit
    if not check_rate_limit(api_key):
        log('WARN', 'Rate limit exceeded', api_key_prefix=api_key[:8] + '...')
        return response(429, {'error': 'Too many requests', 'message': 'Rate limit exceeded'})

    try:
        # GET /responses/pending
        if method == 'GET' and path == '/responses/pending':
            log('AUDIT', 'Listing pending responses')
            return list_pending_responses()

        # GET /responses/all
        if method == 'GET' and path == '/responses/all':
            log('AUDIT', 'Listing all responses')
            return list_all_responses()

        # GET /responses/{id}/download
        if method == 'GET' and path.startswith('/responses/') and path.endswith('/download'):
            parts = path.split('/')
            if len(parts) >= 3:
                response_id = parts[2]
                log('AUDIT', 'Downloading response', id=response_id)
                return download_and_mark_response(response_id)

        # POST /responses/{id}/unmark
        if method == 'POST' and path.startswith('/responses/') and path.endswith('/unmark'):
            parts = path.split('/')
            if len(parts) >= 3:
                response_id = parts[2]
                log('AUDIT', 'Unmarking response', id=response_id)
                return unmark_response(response_id)

        log('WARN', 'Route not found', path=path, method=method)
        return response(404, {'error': 'Not found'})

    except Exception as e:
        log('ERROR', 'Handler error', error=str(e))
        return response(500, {'error': 'Internal server error'})
